### PR TITLE
Merge changes from lspower fork

### PR DESCRIPTION
**Pull request by silvanshade**
This PR follows the discussion in #308

### CODE REVIEW

File: .github/workflows/main.yml
Comment (ebkalderon) [inline .github/workflows/main.yml:91]: Should this key not point to `${{ matrix.rust-version }}` instead of `${{ env.RUST_TOOLCHAIN }}`? Also, I believe `RUST_TOOLCHAIN: stable` should be removed if a build matrix is already in use.

File: .github/workflows/main.yml
Comment (ebkalderon) [inline .github/workflows/main.yml:81]: This `include:` is missing the `continue-on-error: true` setting. Without this, the `include:` statement is useless.

The intention of `continue-on-error` in the original GitHub Workflow is to allow errors from the `nightly` channel to not block merging, but to act purely as a notification of upcoming breakage. It can also serve as a source of bug reports for `rust-lang/rust` itself.

File: .github/workflows/main.yml
Comment (ebkalderon) [inline .github/workflows/main.yml:118]: Same feedback here in `cargo-test-tokio` as in `cargo-test-agnostic` above, regarding the missing `continue-on-error` and removal of `env: RUST_TOOLCHAIN`.

File: src/codec.rs
Comment (ebkalderon) [inline src/codec.rs:113]: Since `trace!()` is already imported, we might want to remove the `log::` prefix to eliminate warnings.

**Comment by silvanshade** [comment]
@ebkalderon I think this is ready for review. One thing to mention is that I rewrote the CI workflow based on the `lspower` one, which is significantly changed, although I did not include all of the changes from that yet (which include coverage and doc generation and some other things). I hope that's okay.

File: src/codec.rs
Comment (ebkalderon) [inline src/codec.rs:198]: I believe this `for`-`if` combo could be replaced with an `if let`-`find` instead? We only ever care about the `Content-Length` header, so there's no reason to keep scanning through the headers if there are more than one.

```suggestion
                    // If the "Content-Length" header exists, parse the value as a `usize`.
                    if let Some(header) = headers.iter().find(|h| h.name == "Content-Length") {
                        match std::str::from_utf8(header.value) {
                            Ok(content_len) => match content_len.parse() {
                                // Successfully set `content_len` from the parsed "Content-Length"
                                // value.
                                Ok(len) => self.content_len = Some(content_len),
                                // If there was an error parsing the "Content-Length" UTF-8 as a
                                // `usize`, set the error.
                                Err(_) => parse = Err(ParseError::InvalidLength),
                            }
                            // If there was an error parsing the "Content-Length" value as UTF-8,
                            // set the error.
                            Err(err) => parse = Err(ParseError::Utf8(err),
                        }
                    }
```

File: src/codec.rs
Comment (ebkalderon) [inline src/codec.rs:233]: Since `trace!()` is already imported, we might want to remove the `log::` prefix to eliminate warnings.

File: src/codec.rs
Comment (ebkalderon) [inline src/codec.rs:222]: Perhaps a clearer name for this might be `message_len` instead of `delta`?

**Comment by silvanshade** [comment]
I've also included (in commit https://github.com/ebkalderon/tower-lsp/pull/309/commits/6cac1aab3ab990045987e3ff85b0dd204d4174a5) changes from this PR: https://github.com/silvanshade/lspower/pull/24

File: src/codec.rs
Comment (ebkalderon) [inline src/codec.rs:157]: Since we already have `std::fmt::Display` imported from up above, perhaps we could also import `Debug` as well for consistency?

File: src/codec.rs
Comment (ebkalderon) [inline src/codec.rs:159]: Very small nitpick, but perhaps we could keep the same ordering of associated types as the upstream definition of the `Decoder` trait, for consistency?

File: src/client.rs
Comment (ebkalderon) [inline src/client.rs:60]: I don't think we want to expose this method to downstream users. Ideally, the `exit` notification alone should dictate when communications between the server and client shut down. I understand this method is only really used in the codegen output of `tower-lsp-macros`. Perhaps we could consider changing the visibility to `pub(crate)` instead to avoid exposing this in the public API?

File: src/codec.rs
Comment (ebkalderon) [inline src/codec.rs:109]: Very small nitpick, but perhaps we could keep the same ordering of associated types as the upstream definition of the `Encoder` trait, for consistency?

**Review by ebkalderon — CHANGES_REQUESTED** [review]
Awesome work here, @silvanshade! Excited to see these changes come through. :heart:

**Comment by ebkalderon** [comment]
Looks like the failure of `cargo-audit` isn't our fault. Rather, there seems to have been an actual outage of crates.io and docs.rs at the moment. I happened to notice this when navigating in my browser. :sweat_smile:

**Comment by silvanshade** [comment]
Thanks for the feedback! All of your suggestions seem reasonable to me. I'll make the appropriate changes.

File: .github/workflows/main.yml
Comment (silvanshade) [reply .github/workflows/main.yml:124]: I think the name (re: "without docs") must have been a typo, probably from copying and pasting parts of the job from another workspace. I agree that `--workspace` doesn't skip the doc tests (and we don't actually want to skip those).

File: .github/workflows/main.yml
Comment (silvanshade) [reply .github/workflows/main.yml:20]: Oh, this is because it allows for either a space or a dash for each variant.

To be honest though, I don't think I've ever actually even used this job. I just keep it around in most of my workflows out of habit. 

I think I'll just remove it unless there's a need for it at some point.

File: src/codec.rs
Comment (silvanshade) [reply src/codec.rs:157]: Oh, whoops, that was actually something I had in place for testing that I forgot to remove.

**Comment by silvanshade** [comment]
@ebkalderon I believe I've addressed all of your feedback with this latest push.

File: .github/workflows/main.yml
Comment (ebkalderon) [reply .github/workflows/main.yml]: Oh wow, I feel dumb! I don't think I noticed the space vs. hyphenated distinction amidst my review! Thanks for pointing it out. :joy:

Comment (ebkalderon) [reply .github/workflows/main.yml]: Oh wow, I feel dumb! I don't think I noticed the space vs. hyphenated distinction during my review! Thanks for pointing out the distinction. I don't mind its presence either way, but it's nice to know this exists. :joy:

File: src/codec.rs
Comment (ebkalderon) [reply src/codec.rs:159]: I see that the associated type ordering in the `Decoder` impl hasn't been addressed yet?

File: src/codec.rs
Comment (ebkalderon) [reply src/codec.rs:157]: No worries! I do see that the `std::fmt::Debug` nit hasn't been addressed yet, though?

File: .github/workflows/main.yml
Comment (ebkalderon) [inline .github/workflows/main.yml:71]: It appears that the `continue-on-error` setting defined in both `cargo-test-agnostic` and `cargo-test-tokio` aren't actually being used in the workflow. This setting needs to be hooked up into _all_ the relevant CI steps by adding:

```diff
 - name: Run cargo test
   uses: actions-rs/cargo@v1
+  continue-on-error: ${{ matrix.continue-on-error || false }}
   with:
     command: test
     args: --lib --no-default-features --features runtime-agnostic
```

Please see the old `rust.yml` workflow for details how this works and on which jobs it should probably be used.

File: .github/workflows/main.yml
Comment (ebkalderon) [inline .github/workflows/main.yml:111]: Any reason why we're using `--workspace` to test all crates in the workspace when testing with the `runtime-tokio` feature and only `--lib` when testing with the `runtime-agnostic` feature? I would imagine we would want the equal test coverage in CI for both Tokio and alternative executors, correct?

**Review by ebkalderon — CHANGES_REQUESTED** [review]
Thanks so much for addressing the feedback! It looks great so far. :sunglasses: 

There do still seem to be a few minor points that appear to be left somewhat unaddressed, so I outlined them in a few comments. I think this PR should be pretty much good to merge after these changes are made, though!

File: .github/workflows/main.yml
Comment (silvanshade) [reply .github/workflows/main.yml:111]: The main reason for this was because the examples currently only compile with the `runtime-tokio` feature. This could probably be fixed though.

File: src/codec.rs
Comment (silvanshade) [reply src/codec.rs:157]: Ah, an oversight. Fixed now.

File: .github/workflows/main.yml
Comment (silvanshade) [reply .github/workflows/main.yml:71]: Okay, I think I see how it works now. Updated the workflow.

**Comment by silvanshade** [comment]
> There do still seem to be a few minor points that appear to be left somewhat unaddressed, so I outlined them in a few comments.

@ebkalderon I think these should be fixed now.

File: .github/workflows/main.yml
Comment (ebkalderon) [reply .github/workflows/main.yml:71]: Sweet! I think this should work great. Thanks so much! :heart:

**Comment by silvanshade** [comment]
@ebkalderon I think the errors from the doc tests should be fixed now.

**Comment by silvanshade** [comment]
I've added a `clippy` job to the CI. I think the error is only because the clippy UI integration doesn't work coming from the forked repo but it should be fine if merged here.

**Review by ebkalderon — APPROVED** [review]
Fantastic work, @silvanshade! Should be ready to merge.

**Pull request merged**
